File: frontend/app/lib/dal/migration.ts
import "server-only";

import {
  type GitHubInstallationListResponse,
  GitHubInstallationListResponseSchema,
  type GitHubInstallationResponse,
  GitHubInstallationResponseSchema,
  type GitHubRepositoryListResponse,
  GitHubRepositoryListResponseSchema,
  type MigrationListResponse,
  MigrationListResponseSchema,
  type MigrationResponse,
  MigrationResponseSchema,
} from "../dto/migration";
import { authFetch, authPost, GITDOT_SERVER_URL, handleResponse } from "./util";

export async function listInstallations(): Promise<GitHubInstallationListResponse | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/migration/github/installations`,
  );

  return await handleResponse(response, GitHubInstallationListResponseSchema);
}

export async function createInstallation(
  installationId: number,
): Promise<GitHubInstallationResponse | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/migration/github/${installationId}`,
    {},
  );

  return await handleResponse(response, GitHubInstallationResponseSchema);
}

export async function listInstallationRepositories(
  installationId: number,
): Promise<GitHubRepositoryListResponse | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/migration/github/${installationId}/repositories`,
  );

  return await handleResponse(response, GitHubRepositoryListResponseSchema);
}

export async function getMigration(
  number: number,
): Promise<MigrationResponse | null> {
  const response = await authFetch(
    `${GITDOT_SERVER_URL}/migration/${number}`,
  );

  return await handleResponse(response, MigrationResponseSchema);
}

export async function listMigrations(): Promise<MigrationListResponse | null> {
  const response = await authFetch(`${GITDOT_SERVER_URL}/migrations`);

  return await handleResponse(response, MigrationListResponseSchema);
}

export async function migrateGitHubRepositories(
  installationId: number,
  owner: string,
  ownerType: string,
  repositories: string[],
): Promise<MigrationResponse | null> {
  const response = await authPost(
    `${GITDOT_SERVER_URL}/migration/github/${installationId}/migrate`,
    { owner, owner_type: ownerType, repositories },
  );

  return await handleResponse(response, MigrationResponseSchema);
}

File: backend/src/handler/migration.rs
mod create_github_installation;
mod get_migration;
mod list_github_installation_repositories;
mod list_github_installations;
mod list_migrations;
mod migrate_github_repositories;

use axum::{
    Router,
    routing::{get, post},
};

use crate::app::AppState;

use create_github_installation::create_github_installation;
use get_migration::get_migration;
use list_github_installation_repositories::list_github_installation_repositories;
use list_github_installations::list_github_installations;
use list_migrations::list_migrations;
use migrate_github_repositories::migrate_github_repositories;

pub fn create_migration_router() -> Router<AppState> {
    Router::new()
        .route("/migrations", get(list_migrations))
        .route("/migration/{number}", get(get_migration))
        .route(
            "/migration/github/installations",
            get(list_github_installations),
        )
        .route(
            "/migration/github/{installation_id}",
            post(create_github_installation),
        )
        .route(
            "/migration/github/{installation_id}/repositories",
            get(list_github_installation_repositories),
        )
        .route(
            "/migration/github/{installation_id}/migrate",
            post(migrate_github_repositories),
        )
}
